pkg/controllers/management/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -441,55 +497,83 @@ func (m *MCAppController) deleteApps(mcAppName string, mcapp *v3.MultiClusterApp
 				return nil, err
 			}
 			appsToDelete = append(appsToDelete, apps...)
+			projectToCluster[projectNS] = split[0]
 		}
 	}
-
-	if err := m.delete(appsToDelete); err != nil {
+	if err := m.delete(appsToDelete, clusterContexts, projectToCluster); err != nil {
 		return nil, err
 	}
 	return nil, nil
 }
 
-func (m *MCAppController) getAllApps(mcAppName string, all bool) ([]*pv3.App, error) {
+func (m *MCAppManager) getAllAppsWithContexts(mcAppName string) ([]*pv3.App, map[string]string, map[string]*config.UserContext, error) {
 	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
 	appsToDelete := []*pv3.App{}
-	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcAppName})
+	projectToCluster := map[string]string{}
+	clusterContexts := map[string]*config.UserContext{}
```

Comment:
lets minimize a number call to clustercontexts.

Reply:
Confirmed, it gets called only twice 
- when mcApp gets deleted and object is nil, there's need to fetch all apps in the deleteApp logic
- on the very start of each sync